### PR TITLE
Singular route request

### PR DESCRIPTION
**Pull request by calebmer**
Continuation of #361 against master.

Removed the single row routes, kept the new router (which we should talk about, do we want it at all?) and implemented @begriffs single object prefer header proposal.

### CODE REVIEW

**Comment by diogob** [comment]
I like the idea of extracting the router to a separate module. Specially since the App module is so big right now. But I'd like to see how the code would look like if we move the whole ```case verb of``` to the Router.

My suggestion is that we create another datatype called Route, which would have and endpoint and a verb.
Then in the App we can pattern match only against the Route instead of having nested cases.

Another suggestion is to add haddock strings for new modules describing their planned functionality, this would help discussing the code and reasoning about what each module should encapsulate.

File: src/PostgREST/App.hs [hunk 1]
```diff
@@ -64,138 +65,154 @@ import           PostgREST.QueryBuilder ( asJson
...
-    ([table], "OPTIONS") -> do
-      let cols = filter (filterCol schema table) allCols
-          pkeys = map pkName $ filter (filterPk schema table) allPrKeys
-          body = encode (TableOptions cols pkeys)
-      return $ responseLBS status200 [jsonH, allOrigins] $ cs body
-
-    (_, _) ->
-      return $ responseLBS status404 [] ""
+  case route RootE dbStructure schema path of
+    Just endpoint ->
```
Comment (begriffs) [inline src/PostgREST/App.hs]: Could decrease the indentation by doing something like

```haskell
  case route RootE dbStructure schema path of
    Nothing -> notFound
    Just RootE ->
      -- ...
    Just (TableE table) ->
      -- ...
```

File: src/PostgREST/App.hs [hunk 2]
```diff
@@ -363,8 +380,8 @@ instance ToJSON TableOptions where
       "columns" .= tblOptcolumns t
     , "pkey"   .= tblOptpkey t ]
 
-parseRequest :: Schema -> [Relation] -> NodeName -> Request -> BL.ByteString -> Either Text (Text, Text, Bool)
-parseRequest schema allRels rootTableName httpRequest reqBody =
+parseRequest :: Schema -> [Relation] -> NodeName -> Request -> [Filter] -> BL.ByteString -> Either Text (Text, Text, Bool)
+parseRequest schema allRels rootTableName httpRequest extraFilters reqBody =
```
Comment (ruslantalpa) [inline src/PostgREST/App.hs]: where are extraFilters used? i think it's a leftover form single route needs

**Comment by ruslantalpa** [comment]
This PR again introduces lots of other things (extrafilters, rewrite of createStatement, dbFindTable, order by, pgErrResponse) which have noting to do with route functionality.
How about closing this PR and adding the router functionality in a clean branch (baed on current master)

**Comment by ruslantalpa** [comment]
Writing of the "Router" data type should be easier if it's based on the refactoring from #371

**Comment by calebmer** [comment]
@ruslantalpa `extraFilters` is leftover, `createStatement` rewrite adds no code (except for some `isSimgle` stuff) `dbFindTable` could be moved to Router.hs, and `pgErrRespinse` is a needed refactor of non-uniform error handling (some errors were just plaintext strings).

The rest of the PR has a rather low footprint, but if we want to merge #371 first that's fine. I'll just do a rebase.

**Comment by begriffs** [comment]
Other refactor merged. Ready for a rebase of this one to add plurality selection.

**Comment by calebmer** [comment]
Do we want Router.hs? I could rebase or split this into two commits. One for the error handling refactor and another which adds plurality singular.

**Comment by PierreR** [comment]
Is there any reason not to use Servant for `routing` ?

**Comment by begriffs** [comment]
I like dealing with small focused PRs best. Splitting it would be most convenient.

**Comment by calebmer** [comment]
@diogob I like that idea too, however it seems for now we are going to pass on implementing a router until we need one.

Closing this PR in favor of two smaller ones: #374 and #375.

**Pull request closed without merging**